### DISCUSSION TITLE
Inefficient error handling on JsonDeserializer

### DISCUSSION THREAD

**Discussion by ccrafael** [spring-projects/spring-kafka#2837] [Ideas]
Actually in the class org.springframework.kafka.support.serializer.JsonDeserializer when an error happen it is handled like this:

```
catch (IOException e) {
		throw new SerializationException("Can't deserialize data [" + Arrays.toString(data) +
				"] from topic [" + topic + "]", e);
	}
```
which I think is very inefficient as each byte of the payload is duplicated as a long array of numbers written as a String in the exception message. In my humble opinion the Arrays.toString(data) should be removed from the message and the original data[] object should be added to the SerializationException class.

**Comment by garyrussell** [comment]
Good point, but `SerializationException` is not part of this project, but we could make a sub class.

However, adding the `byte[]` to the  to the exception is moot - because, normally, this is used in conjunction with an `ErrorHandlingDerserializer`, which adds the raw data to the `org.springframework.kafka.support.serializer.DeserializationException` anyway.

Not using an `ErrorHandlingDerserializer` is pointless because Spring can't recover from such situations (and `kafka-clients` is unaware).

I will convert this to an issue; to remove the `Arrays.toString()`.

**Reply by garyrussell** [reply]
https://github.com/spring-projects/spring-kafka/issues/2838

**Reply by garyrussell** [reply]
However, when NOT using an EHD, the user will have no clue as to why the deserialization failed (which was probably the original reason that we included the complete JSON in the exception). So maybe we should retain it in that case.

**Reply by garyrussell** [reply]
On the other hand, Jackson does report the offset within the data where the problem occurred, so the user could get the data using the command line tools.

**Reply by ccrafael** [reply]
d'oh!! I didnt pay attention! You are right SerializationException  is a kafka class. Anyway ErrorHandlingDerserializer is actually wrapping the exception into a DeserializationException, adding the data, the headers and serializing all using Java so the error with the data  in the message already duplicated is also serialized and the data is duplicated in memory again.

**Reply by garyrussell** [reply]
Regardless of the duplication, I agree the array of numbers was useless; so I have removed it.

https://docs.spring.io/spring-kafka/reference/whats-new.html

(also back ported to 3.0.x, 2.9.x).
